### DISCUSSION TITLE
Ops in Parallel

### DISCUSSION THREAD

**Discussion by championmeatball** [dagster-io/dagster#14546] [Q&A] [answered]
First of all, thanks in advance for any help that is provided here.  I'm new to Dagster and Python so fumbling a bit through the dark while I figure this out. 

I am trying to figure out how to run an op in parallel based of a first op.  I have one op that generates a list of file names and then a second op which takes a given file name, pulls that file from S3, parses the file, and writes the results to my DB. The second op can be run in parallel.

As example I am trying:

```
@op()
def my_first_op() -> List[str]:
    return list_of_file 
```
```
@op()
def my_second_op(file):
    # process file
```

```
@graph
def process_my_files():
    file_list = my_first_op()
    for file in file_list:
        my_second_op(file)
```

but I get

> dagster._core.errors.DagsterInvariantViolationError: Attempted to iterate over an InvokedNodeOutputHandle. This object represents the output "result" from the op/graph "my_first_op". Consider defining multiple Outs if you seek to pass different parts of this output to different op/graph.

My list of files is quite large here (over 1 million), so I'm trying to do all the processing parallel as quickly as possible. Is there a way to do this?

Thanks again for the help!

**Comment by clairelin135** [comment, marked as answer]
Hi @championmeatball. This looks like a good use case for dynamic outputs: https://docs.dagster.io/concepts/ops-jobs-graphs/dynamic-graphs#a-dynamic-job

You can yield each file as a dynamic output, and map over these outputs using `my_second_op`. This will allow each mapped instance of `my_second_op` to run in parallel.

**Reply by championmeatball** [reply]
Hi @clairelin135 good info here. Thank you!  I found a very similar thread on the Slack channel (https://dagster.slack.com/archives/C01U954MEER/p1685052366889939?thread_ts=1684980703.770879&cid=C01U954MEER) and am running into the same problem. I'm not saving much time running the ops in parallel because it takes so long to run the loop to produce the dynamic output.  Any suggestions on how to overcome this?

**Reply by akshayv** [reply]
Hi @championmeatball did you figure a way around your issue?
